server/src/main/java/org/opensearch/cluster/block/ClusterBlock.java

```diff
@@ -154,24 +175,99 @@ public boolean disableStatePersistence() {
         }
         builder.endArray();
+        if (context == Metadata.XContentContext.GATEWAY) {
+            builder.field(KEY_STATUS, status);
+            builder.field(KEY_ALLOW_RELEASE_RESOURCES, allowReleaseResources);
+        }
         builder.endObject();
         return builder;
     }
 
+    public static ClusterBlock fromXContent(XContentParser parser, int id) throws IOException {
+        String uuid = null;
+        String description = null;
+        boolean retryable = false;
+        boolean disableStatePersistence = false;
+        RestStatus status = null;
+        boolean allowReleaseResources = false;
+        EnumSet<ClusterBlockLevel> levels = EnumSet.noneOf(ClusterBlockLevel.class);
+        String currentFieldName = skipBlockID(parser);
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                switch (Objects.requireNonNull(currentFieldName)) {
+                    case KEY_UUID:
+                        uuid = parser.text();
+                        break;
+                    case KEY_DESCRIPTION:
+                        description = parser.text();
+                        break;
+                    case KEY_RETRYABLE:
+                        retryable = parser.booleanValue();
+                        break;
+                    case KEY_DISABLE_STATE_PERSISTENCE:
+                        disableStatePersistence = parser.booleanValue();
+                        break;
+                    case KEY_STATUS:
+                        status = RestStatus.valueOf(parser.text());
+                        break;
+                    case KEY_ALLOW_RELEASE_RESOURCES:
+                        allowReleaseResources = parser.booleanValue();
+                        break;
+                    default:
+                        throw new IllegalArgumentException("unknown field [" + currentFieldName + "]");
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                if (currentFieldName.equals(KEY_LEVELS)) {
+                    while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                        levels.add(ClusterBlockLevel.fromString(parser.text(), Locale.ROOT));
+                    }
+                } else {
+                    throw new IllegalArgumentException("unknown field [" + currentFieldName + "]");
+                }
+            } else {
+                throw new IllegalArgumentException("unexpected token [" + token + "]");
+            }
+        }
+        return new ClusterBlock(id, uuid, description, retryable, disableStatePersistence, allowReleaseResources, status, levels);
+    }
+
+    private static String skipBlockID(XContentParser parser) throws IOException {
```

Comment:
Can we add explanation of why this is needed? Probably an example will help here

Reply:
How do we differentiate between XContentObject and XConentFragment? Can we separate XContentObject parsing to a separate function and then use the above method to parse the XConentFragment inside that?

Reply:
XContentObject has a paired start end object. While for fragment that is not true. Sure, we can separate the logic to parse the fragment to a `innerParser` and use that in `fromXContent` method which can accept both Object and a Fragment.